.github/actions/sonar/action.yml

```
@@ -0,0 +1,268 @@
+    required: true
+
+runs:
+  using: composite
+  steps:
+    # Make sure the sonar project folder exists.
+    - run: mkdir -p ./${{ inputs.sonar-project-key }}
+      shell: bash
+
+    # Cache the main branch project for future PR analysis
+    - name: 'Cache project on main branch'
+      if: github.ref == 'refs/heads/main'
+      shell: bash
+      run: |
+        cp -r ${{ inputs.workdir }}/. ./${{ inputs.sonar-project-key }}/
+        rm -rf ${{ inputs.sonar-project-key }}/node_modules \
+            ${{ inputs.sonar-project-key }}/package-lock.json \
+            ${{ inputs.sonar-project-key }}/target/node \
+            ${{ inputs.sonar-project-key }}/target/*.original \
+            ${{ inputs.sonar-project-key }}/target/*.jar
+
+    - name: 'Store cache (main branch)'
+      if: github.ref == 'refs/heads/main'
+      uses: actions/cache/save@v4
+      with:
+        path: ./${{ inputs.sonar-project-key }}
+        key: application-${{ inputs.sonar-project-key }}-${{ github.sha }}
+
+    # Restore the cache for PR analysis
+    - name: 'Restore project cache for PR analysis'
+      if: github.event_name == 'pull_request'
+      id: restore_cache
+      uses: actions/cache/restore@v4
+      with:
+        path: ./${{ inputs.sonar-project-key }}
+        key: application-${{ inputs.sonar-project-key }}-${{ github.event.pull_request.base.sha }}
+        restore-keys: |
+          application-${{ inputs.sonar-project-key }}-
+
+    # Start SonarQube server using Docker Compose
+    - name: 'Start SonarQube server'
+      if: github.event_name == 'pull_request'
+      shell: bash
+      run: docker compose -f ${{ inputs.docker-compose-file }} up --build -d --wait
+
+    # Create SonarQube project
+    - name: 'Create SonarQube project'
+      if: github.event_name == 'pull_request'
+      shell: bash
+      run: |
+        curl -s -u admin:admin -X POST "http://localhost:9000/api/projects/create?name=${{ inputs.sonar-project-key }}&project=${{ inputs.sonar-project-key }}" || true
+
+    # Run SonarQube analysis on the main branch or add an empty scan
+    - name: 'Run SonarQube analysis on main branch'
+      if: github.event_name == 'pull_request' && steps.restore_cache.outputs.cache-hit == 'true'
+      shell: bash
+      run: >-
+        ./mvnw --batch-mode initialize org.jacoco:jacoco-maven-plugin:prepare-agent sonar:sonar
+        -Dsonar.host.url=http://localhost:9000
+        -Dsonar.projectKey=${{ inputs.sonar-project-key }}
+        -Dsonar.login=admin
+        -Dsonar.password=admin
+        -Dsonar.branch.name=main
+      working-directory: ${{ inputs.sonar-project-key }}
+
+    - name: Add an empty scan to sonar main branch
+      if: github.event_name == 'pull_request' && steps.restore_cache.outputs.cache-hit != 'true'
+      shell: bash
+      run: >-
+        git init &&
+        git commit -m "Initial commit" --allow-empty &&
+        docker run --net=host --rm sonarsource/sonar-scanner-cli
```

Add volume so git is detected.
```suggestion
        docker run --net=host -v ".:/usr/src" --rm sonarsource/sonar-scanner-cli
```